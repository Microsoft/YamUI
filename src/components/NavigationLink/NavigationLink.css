/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
.y-navigationLink__unstyled,
.y-navigationLink__unstyled:active,
.y-navigationLink__unstyled:focus,
.y-navigationLink__unstyled:hover {
  color: inherit;
  cursor: pointer;
  text-decoration: none;
}
