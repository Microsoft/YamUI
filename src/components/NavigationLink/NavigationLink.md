### Notes for use

`NavigationLink` purposefully does not offer an `onClick` event as it should only be used for navigation. Please use a `Clickable` component for event handlers which do not change the current URL.

### Examples

Basic:

```js { "props": { "data-description": "basic", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-navigationLink\"},{\"action\":\"focus\",\"selector\":\".y-navigationLink\"}]" } }
<div>
  Here is a <NavigationLink href="/404.html"><Strong>NavigationLink</Strong> component</NavigationLink> with just an <Strong>href</Strong>, its only required property.
</div>
```

Opens a new window:

```js { "props": { "data-description": "new window" } }
<div>
  Here is a <NavigationLink href="/404.html" newWindow={true}><Strong>NavigationLink</Strong> component</NavigationLink> which opens its link securely in a new window.
</div>
```

Unstyled link:

```js { "props": { "data-description": "unstyled" , "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-navigationLink\"},{\"action\":\"focus\",\"selector\":\".y-navigationLink\"}]" } }
<div>
  Here is a <NavigationLink href="/404.html" unstyled={true}><Strong>NavigationLink</Strong> component</NavigationLink> which removes the link styling and inherits its parent's color.
</div>
```

Link wrapping content:

```js { "props": { "data-description": "wrapping", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-navigationLink\"},{\"action\":\"focus\",\"selector\":\".y-navigationLink\"}]" } }
<NavigationLink href="/404.html" block={true}>
  <div>
    This <Strong>NavigationLink</Strong> component wraps an entire block of text. By default all text content within a <Strong>NavigationLink</Strong> tag will be styled like a link.
  </div>
</NavigationLink>
```

Unstyled link wrapping content:

```js { "props": { "data-description": "unstyled wrapping", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-navigationLink\"},{\"action\":\"focus\",\"selector\":\".y-navigationLink\"}]" } }
const { TextColor } = require('../Text');

<div>
  <NavigationLink href="/404.html" unstyled={true} block={true}>
    <div>
      This is an unstyled <Strong>NavigationLink</Strong> component wrapping an entire block of text.
    </div>
  </NavigationLink>
  <br />
  <NavigationLink href="/404.html" unstyled={true} block={true}>
    <div>
      <Text color={TextColor.METADATA}>
        This is another unstyled <Strong>NavigationLink</Strong> component wrapping a block of text with a different color.
      </Text>
    </div>
  </NavigationLink>
</div>
```

Wrapping image:

```js { "props": { "data-description": "wrapping image" } }
const yammerLogo = 'logo.png';
const yammerLogoDescription = 'Yammer "y" logo';

<div>
  <NavigationLink href="/404.html" unstyled={true} block={true}>
    <Image source={yammerLogo} description={yammerLogoDescription} width={320} height={240} block={true} />
  </NavigationLink>
</div>
```
