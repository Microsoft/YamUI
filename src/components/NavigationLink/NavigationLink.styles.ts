/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { mergeStyleSets } from '@uifabric/styling';
import { memoizeFunction } from '@uifabric/utilities';

export interface NavigationLinkStyleProps {
  block?: boolean;
  unstyled?: boolean;
}

export const getClassNames = memoizeFunction((styleProps: NavigationLinkStyleProps) => {
  const { unstyled, block } = styleProps;

  const overrides = unstyled
    ? {
        color: 'inherit',
        cursor: 'pointer',
        textDecoration: 'none',
      }
    : {};

  return mergeStyleSets({
    root: {
      display: block ? 'block' : undefined,
      ...overrides,
      selectors: {
        ':active': overrides,
        ':focus': overrides,
        ':hover': overrides,
      },
    },
  });
});
