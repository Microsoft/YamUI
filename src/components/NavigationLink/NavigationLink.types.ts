/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';

export interface NavigationLinkProps extends NestableBaseComponentProps {
  /**
   * Whether the clickable should be `display: block`.
   */
  block?: boolean;

  /**
   * URL or a URL fragment that the link points to.
   */
  href: string;

  /**
   * Whether the link should open in a new window. It securely removes access to the opening window.
   */
  newWindow?: boolean;

  /**
   * If present, indicates the link refers to the current page. This is usually used in cases where the link
   * is visually styled to indicate it is selected or is the current page, such as in a navigation list.
   */
  ariaCurrent?: 'page';

  /**
   * Title or description of the linked document for screenreaders.
   */
  ariaLabel?: string;

  /**
   * Title or description of the linked document for the anchor tag. You should generally avoid using this.
   */
  title?: string;

  /**
   * Whether to remove all styles from the link. Useful for allowing a large area to be clickable
   * while nesting `FakeLink` components to show link and hover state visuals.
   */
  unstyled?: boolean;
}
