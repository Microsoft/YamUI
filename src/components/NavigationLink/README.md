# NavigationLink

This renders an `<a>` tag for navigation between web pages.

## Properties

| Prop | Type | Description |
| ---- | ---- | ----------- |
| href * | string | The URL |
| newWindow | boolean | If true, will open the URL in a new window and securely break any reference back to this window. |
| unstyled | boolean | If true, will render the content as if it were not nested in an `<a>` tag. |

## Notes for use

`NavigationLink` purposefully does not offer an `onClick` event as it should only be used for navigation. Please use a `Clickable` component for event handlers.

## Examples

### Basic usage

```js
<NavigationLink href="my_page.html">
  This NavigationLink is tabbable and navigates to the given href when clicked.
</NavigationLink>.
```
