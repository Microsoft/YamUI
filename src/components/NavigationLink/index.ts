/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export { default } from './NavigationLink';
export * from './NavigationLink.types';
