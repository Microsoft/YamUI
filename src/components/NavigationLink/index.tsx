/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import classNames = require('classnames');
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import { secureOpen } from '../../util/secureOpener';
import './navigation-link.css';


export interface NavigationLinkProps extends NestableBaseComponentProps {
  href: string;
  title?: string;
  newWindow?: boolean;
  unstyled?: boolean;
}

export default class NavigationLink extends React.Component<NavigationLinkProps, {}> {
  constructor (props: NavigationLinkProps) {
    super(props);
    this.handleClick = this.handleClick.bind(this);
  }

  getClasses () {
    const classes: string[] = [
      'y-navigationLink',
    ];
    if (this.props.newWindow) {
      classes.push('y-navigationLink__newWindow');
    }
    if (this.props.unstyled) {
      classes.push('y-navigationLink__unstyled');
    }
    if (this.props.className) {
      classes.push(this.props.className);
    }
    return classNames(classes);
  }

  handleClick (event: React.SyntheticEvent<HTMLAnchorElement>) {
    if (!this.props.newWindow) {
      return;
    }

    secureOpen(this.props.href);
    event.preventDefault();
  }

  render () {
    let target;
    let rel;
    if (this.props.newWindow) {
      target = '_blank';
      rel = 'nofollow noreferrer';
    }
    return (
      <a href={this.props.href} className={this.getClasses()} onClick={this.handleClick}
         target={target} rel={rel} title={this.props.title}>
        {this.props.children}
      </a>
    );
  }
}
