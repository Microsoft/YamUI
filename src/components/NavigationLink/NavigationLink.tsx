/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { CustomizableComponentProps, defaultTheme, customizable } from '../Customizer';
import { join } from '../../util/classNames';
import { getClassNames } from './NavigationLink.styles';
import { NavigationLinkProps } from './NavigationLink.types';

/**
 * A `NavigationLink` renders an `a` tag for navigation between web pages.
 */
export class NavigationLink extends React.Component<NavigationLinkProps & CustomizableComponentProps> {
  public render() {
    const {
      ariaCurrent,
      ariaLabel,
      href,
      newWindow,
      title,
      children,
      unstyled,
      block,
      className,
      theme = defaultTheme,
    } = this.props;
    const classNames = getClassNames({ unstyled, block, theme });

    return (
      <a
        className={join(['y-navigationLink', className, classNames.root])}
        href={href}
        rel={newWindow ? 'nofollow noopener noreferrer' : undefined}
        target={newWindow ? '_blank' : undefined}
        title={title}
        aria-label={ariaLabel}
        aria-current={ariaCurrent}
      >
        {children}
      </a>
    );
  }
}

@customizable('NavigationLink', ['theme'])
export default class CustomizableNavigationLink extends NavigationLink {}
