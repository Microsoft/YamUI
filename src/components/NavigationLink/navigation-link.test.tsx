/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import NavigationLink, { NavigationLinkProps } from './index';
import * as secureOpener from '../../util/secureOpener';

describe('<NavigationLink />', () => {
  let component: ShallowWrapper<NavigationLinkProps, {}>;

  describe('with default options', () => {
    beforeEach(() => {
      component = shallow(<NavigationLink href="test.html">link content</NavigationLink>);
    });

    it('renders its given content', () => {
      expect(component.text()).toEqual('link content');
    });

    it('contains its base className', () => {
      expect(component.hasClass('y-navigationLink')).toBe(true);
    });

    it('uses the given href', () => {
      expect(component.getNode().props.href).toEqual('test.html');
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    describe('when clicked', () => {
      let spy: jest.SpyInstance<{}>;
      beforeEach(() => {
        spy = jest.spyOn(secureOpener, 'secureOpen');
      });

      it('does not open in a new window', () => {
        component.simulate('click');
        expect(spy).not.toHaveBeenCalled();
      });
    });
  });

  describe('when unstyled', () => {
    beforeEach(() => {
      component = shallow(
        <NavigationLink href="test.html" unstyled={true}>
          link content
        </NavigationLink>,
      );
    });

    it('contains its base className', () => {
      expect(component.hasClass('y-navigationLink')).toBe(true);
    });

    it('contains its unstyled className', () => {
      expect(component.hasClass('y-navigationLink__unstyled')).toBe(true);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with title', () => {
    beforeEach(() => {
      component = shallow(
        <NavigationLink href="test.html" title="TITLE CONTENT">
          link content
        </NavigationLink>,
      );
    });

    it('contains the given title', () => {
      expect(component.getNode().props.title).toEqual('TITLE CONTENT');
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('when newWindow is true', () => {
    beforeEach(() => {
      component = shallow(
        <NavigationLink href="test.html" newWindow={true}>
          link content
        </NavigationLink>,
      );
    });

    it('contains its newWindow className', () => {
      expect(component.hasClass('y-navigationLink__newWindow')).toBe(true);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    describe('when clicked', () => {
      let spy: jest.SpyInstance<{}>;
      beforeEach(() => {
        spy = jest.spyOn(secureOpener, 'secureOpen');
      });

      it('opens in a new window', () => {
        component.simulate('click', { preventDefault: () => undefined });
        expect(spy).toHaveBeenCalled();
      });
    });
  });

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(
        <NavigationLink href="test.html" className="TEST_CLASSNAME">
          test content
        </NavigationLink>,
      );
    });

    it('includes that className', () => {
      expect(component.hasClass('TEST_CLASSNAME')).toBeTruthy();
    });

    it('still includes its base className', () => {
      expect(component.hasClass('y-navigationLink')).toBeTruthy();
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
