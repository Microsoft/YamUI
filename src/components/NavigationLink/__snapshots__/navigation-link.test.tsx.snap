// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavigationLink /> when newWindow is true matches its snapshot 1`] = `
<a
  className="y-navigationLink y-navigationLink__newWindow"
  href="test.html"
  onClick={[Function]}
  rel="nofollow noreferrer"
  target="_blank"
>
  link content
</a>
`;

exports[`<NavigationLink /> when unstyled matches its snapshot 1`] = `
<a
  className="y-navigationLink y-navigationLink__unstyled"
  href="test.html"
  onClick={[Function]}
>
  link content
</a>
`;

exports[`<NavigationLink /> with additional className matches its snapshot 1`] = `
<a
  className="y-navigationLink TEST_CLASSNAME"
  href="test.html"
  onClick={[Function]}
>
  test content
</a>
`;

exports[`<NavigationLink /> with default options matches its snapshot 1`] = `
<a
  className="y-navigationLink"
  href="test.html"
  onClick={[Function]}
>
  link content
</a>
`;

exports[`<NavigationLink /> with title matches its snapshot 1`] = `
<a
  className="y-navigationLink"
  href="test.html"
  onClick={[Function]}
  title="TITLE CONTENT"
>
  link content
</a>
`;
